File: python/doc_test.py
# tests that correspond to the `src/doc.rs` rust api
import asyncio
import os
import random
import tempfile

import iroh
import pytest
from iroh import (
    AuthorId,
    Iroh,
    NodeAddr,
    PublicKey,
    Query,
    QueryOptions,
    SortBy,
    SortDirection,
    key_to_path,
    path_to_key,
)


def test_node_addr():
    #
    # create a node_id
    key_str = "ki6htfv2252cj2lhq3hxu4qfcfjtpjnukzonevigudzjpmmruxva"
    node_id = PublicKey.from_string(key_str)
    #
    # create socketaddrs
    ipv4 = "127.0.0.1:3000"
    ipv6 = "::1:3000"
    #
    # relay url
    relay_url = "https://example.com"
    #
    # create a NodeAddr
    expect_addrs = [ipv4, ipv6]
    node_addr = NodeAddr(node_id, relay_url, expect_addrs)
    #
    # test we have returned the expected addresses
    got_addrs = node_addr.direct_addresses()
    for (got, expect) in zip(got_addrs, expect_addrs):
        assert got == expect

    assert relay_url == node_addr.relay_url()

def test_author_id():
    #
    # create id from string
    author_str = "mqtlzayyv4pb4xvnqnw5wxb2meivzq5ze6jihpa7fv5lfwdoya4q"
    author = AuthorId.from_string(author_str)
    #
    # call to_string, ensure equal
    assert str(author) == author_str
    #
    # create another id, same string
    author_0 = AuthorId.from_string(author_str)
    #
    # ensure equal
    assert author.equal(author_0)
    assert author_0.equal(author)

def test_query():
    opts = QueryOptions(sort_by=SortBy.KEY_AUTHOR, direction=SortDirection.ASC, offset=10, limit=10)
    # all
    all = Query.all(opts)
    assert all.offset() == 10
    assert all.limit() == 10

    # single_latest_per_key
    opts.direction = SortDirection.DESC
    opts.limit = 0
    opts.offset = 0
    single_latest_per_key = Query.single_latest_per_key(opts)
    assert single_latest_per_key.offset() == 0
    assert single_latest_per_key.limit() is None

    # author
    opts.direction = SortDirection.ASC
    opts.offset = 100
    author = Query.author(AuthorId.from_string("mqtlzayyv4pb4xvnqnw5wxb2meivzq5ze6jihpa7fv5lfwdoya4q"), opts)
    assert author.offset() == 100
    assert author.limit() is None

    # key_exact
    opts.sort_by = SortBy.KEY_AUTHOR
    opts.direction = SortDirection.DESC
    opts.offset = 0
    opts.limit = 100
    key_exact = Query.key_exact(
        b"key",
        opts,
    )
    assert key_exact.offset() == 0
    assert key_exact.limit() == 100

    # key_prefix
    key_prefix = Query.key_prefix(
        b"prefix",
        opts,
    )
    assert key_prefix.offset() == 0
    assert key_prefix.limit() == 100

@pytest.mark.asyncio()
async def test_doc_entry_basics():
    # setup event loop, to ensure async callbacks work
    iroh.iroh_ffi.uniffi_set_event_loop(asyncio.get_running_loop())

    #
    # create node
    dir = tempfile.TemporaryDirectory()
    node = await Iroh.persistent(dir.name)
    #
    # create doc and author
    doc = await node.docs().create()
    author = await node.authors().create()
    #
    # create entry
    val = b"hello world!"
    key = b"foo"
    hash = await doc.set_bytes(author, key, val)
    #
    # get entry
    query = Query.author_key_exact(author, key)
    entry = await doc.get_one(query)
    assert hash.equal(entry.content_hash())
    assert len(val) == entry.content_len()
    got_val = await entry.content_bytes(doc)
    assert val == got_val

@pytest.mark.asyncio()
async def test_doc_import_export():
    # setup event loop, to ensure async callbacks work
    iroh.iroh_ffi.uniffi_set_event_loop(asyncio.get_running_loop())

    #
    # create file temp der
    dir = tempfile.TemporaryDirectory()
    in_root = os.path.join(dir.name, "in")
    out_root = os.path.join(dir.name, "out")
    os.makedirs(in_root, exist_ok=True)
    os.makedirs(out_root, exist_ok=True)
    #
    # create file
    path = os.path.join(in_root, "test")
    size = 100
    bytes = bytearray(map(random.getrandbits,(8,)*size))
    with open(path, "wb") as file:  # noqa: ASYNC230 blocking-open-call-in-async-function
        file.write(bytes)
    #
    # create node
    iroh_dir = tempfile.TemporaryDirectory()
    node = await Iroh.persistent(iroh_dir.name)
    #
    # create doc and author
    doc = await node.docs().create()
    author = await node.authors().create()
    #
    # import entry
    key = path_to_key(path, None, in_root)
    await doc.import_file(author, key, path, True, None)
    #
    # get entry
    query = Query.author_key_exact(author, key)
    entry = await doc.get_one(query)
    #
    # export entry
    path = key_to_path(key, None, out_root)
    await doc.export_file(entry, path, None)
    #
    # read file
    with open(path, "rb") as file:  # noqa: ASYNC230 blocking-open-call-in-async-function
        got_bytes = file.read()
    assert bytes == got_bytes

File: js/doc.test.js
const { IrohNode, PublicKey, NodeAddr, AuthorId, Query, SortBy, SortDirection, pathToKey, keyToPath } = require( '../index');
const path = require('path');

test('node address', () => {
    const keyStr = "ki6htfv2252cj2lhq3hxu4qfcfjtpjnukzonevigudzjpmmruxva";
    const nodeID = PublicKey.fromString(keyStr);
    const ipv4 = "127.0.0.1:3000";
    const ipv6 = "::1:3000";
    const derpURL = "https://example.com";
    const expectAddrs = [ipv4, ipv6];
    const nodeAddr = new NodeAddr(nodeID, derpURL, expectAddrs);
    const gotAddrs = nodeAddr.directAddresses;
    expect(gotAddrs).toEqual(expectAddrs);
    expect(nodeAddr.derpUrl).toBe(derpURL);
});

test('author ID', () => {
    const authorStr = "mqtlzayyv4pb4xvnqnw5wxb2meivzq5ze6jihpa7fv5lfwdoya4q";
    const author = AuthorId.fromString(authorStr);
    expect(author.toString()).toBe(authorStr);
    const author0 = AuthorId.fromString(authorStr);
    expect(author.equal(author0)).toBe(true);
});

test('query', () => {
    const big10 = BigInt(10);
    const opts = { sortBy: SortBy.KeyAuthor, direction: SortDirection.Asc, offset: big10, limit: big10};
    const all = Query.all(opts);
    expect(all.offset()).toEqual(big10);
    expect(all.limit()).toEqual(big10);

    opts.direction = SortDirection.Desc;
    const big0 = BigInt(0);
    opts.limit = big0;
    opts.offset = big0;
    const singleLatestPerKey = Query.singleLatestPerKey(opts);
    expect(singleLatestPerKey.offset()).toEqual(big0);
    expect(singleLatestPerKey.limit()).toBe(null);

    opts.direction = SortDirection.Asc;
    const big100 = BigInt(100);
    opts.offset = big100;
    const author = Query.author(AuthorId.fromString("mqtlzayyv4pb4xvnqnw5wxb2meivzq5ze6jihpa7fv5lfwdoya4q"), opts);
    expect(author.offset()).toEqual(big100);
    expect(author.limit()).toBe(null);

    opts.sortBy = SortBy.KeyAuthor;
    opts.direction = SortDirection.Desc;
    opts.offset = big0;
    opts.limit = big100;
    const keyExact = Query.keyExact(Buffer.from('key'), opts);
    expect(keyExact.offset()).toEqual(big0);
    expect(keyExact.limit()).toEqual(big100);

    const keyPrefix = Query.keyPrefix(Buffer.from('prefix'), opts);
    expect(keyPrefix.offset()).toEqual(big0);
    expect(keyPrefix.limit()).toEqual(big100);
});

test('document entry basics', async () => {
    const dir = path.join(require('os').tmpdir(), 'document_entry_basics');
    const node = await IrohNode.withPath(dir);
    const author = await node.authorCreate();
    const doc = await node.docCreate();
    const val = Buffer.from('hello world!');
    const key = Buffer.from('foo');
    const hash = await doc.setBytes(author, key, val);
    const query = Query.authorKeyExact(author, key);
    const entry = await doc.getOne(query);
    expect(hash.equal(entry.contentHash())).toBe(true);
    expect(BigInt(val.length)).toEqual(entry.contentLen());
    const gotVal = await entry.contentBytes(doc);
    expect(gotVal).toEqual(val);
});

test('document import export', async () => {
    const tmp = require('os').tmpdir();
    const inRoot = require('path').join(tmp, "in");
    const outRoot = require('path').join(tmp, "out");
    require('fs').mkdirSync(inRoot, { recursive: true });
    require('fs').mkdirSync(outRoot, { recursive: true });
    const path = require('path').join(inRoot, "test");
    const size = 100;
    const bytes = Buffer.alloc(size).map(() => Math.floor(Math.random() * 256));
    require('fs').writeFileSync(path, bytes);
    const irohDir = require('os').tmpdir();
    const node = await IrohNode.withPath(irohDir);
    const doc = await node.docCreate();
    const author = await node.authorCreate();
    const key = pathToKey(path, null, inRoot);
    let importProgress = await doc.importFile(author, key, path, true, null);
    const iter = importProgress[Symbol.iterator]();
    iter.return();
    const query = Query.authorKeyExact(author, key);
    const entry = await doc.getOne(query);
    const outPath = keyToPath(key, null, outRoot);
    let exportProgress = await doc.exportFile(entry, outPath);
    const expIter = exportProgress[Symbol.iterator]();
    expIter.return();
    const gotBytes = require('fs').readFileSync(outPath);
    expect(gotBytes.equals(bytes)).toBe(true);
});
